<a href="https://colab.research.google.com/github/ITouray/CS-448/blob/main/Assignment%203%20-%20Equation%20of%20a%20slime.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://github.com/ITouray/CS-448/blob/main/Assignment%203%20-%20Equation%20of%20a%20slime.ipynb)

In [ ]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn import datasets
from sklearn import svm
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
from sklearn import metrics
from sklearn.metrics import r2_score
from sklearn.model_selection import cross_val_score
from numpy import asarray
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import ShuffleSplit
from IPython.display import display, Math, Latex

# Part 1. Loading the dataset

In [ ]:
# Using pandas load the dataset (load remotely, not locally)
df = pd.read_csv('https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv')


In [ ]:
# Output the first 15 rows of the data
df.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [ ]:
# Display a summary of the table information (number of datapoints, etc.)
len(df)

1000

In [ ]:
df.describe()

Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

# Part 2. Splitting the dataset

In [ ]:
# Take the pandas dataset and split it into our features (X) and label (y)

features = ["Mols KCL", "Size nm^3"]
label = ["Temperature °C"]

In [ ]:
x = df[features].values
y = df[label].values

In [ ]:
# x , y

In [ ]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = .1, random_state = 0)

In [ ]:
x_train.shape , x_test.shape , y_train.shape , y_test.shape

((900, 2), (100, 2), (900, 1), (100, 1))

# Part 3. Perform a Linear Regression

In [ ]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(x_train, y_train)

LinearRegression()

In [ ]:
# Create a sample datapoint and predict the output of that sample with the trained model

y_pred = model.predict(x_test)
y_pred

array([[ 389.61986015],
       [ 575.21612064],
       [1114.44411221],
       [ 249.88117485],
       [ 272.60958484],
       [ 300.17801125],
       [ 203.48481326],
       [ 853.56557591],
       [ 232.20107385],
       [ 281.72863101],
       [ 476.74695771],
       [1065.18995783],
       [ 782.22069437],
       [ 383.68987485],
       [ 229.37542796],
       [1036.51433101],
       [ 258.79486236],
       [ 597.33782993],
       [ 472.34656507],
       [ 551.17261323],
       [1234.08067822],
       [ 410.31480796],
       [ 554.46613542],
       [1211.09824566],
       [ 392.92403359],
       [ 999.17844871],
       [ 215.42657808],
       [ 529.3687714 ],
       [ 529.88816997],
       [ 386.26900228],
       [ 409.44180246],
       [ 145.90250563],
       [1275.30404849],
       [ 603.72326741],
       [1206.32003184],
       [ 434.93207673],
       [ 710.60275556],
       [ 918.95506319],
       [1051.92538328],
       [ 464.51967653],
       [ 204.19578745],
       [ 242.043

In [ ]:
model.predict([[647, 6.244743e+05]])

array([[459.43732195]])

In [ ]:
# predict with r2 score
r2_score(y_test, y_pred)


0.7201592080149567

In [ ]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# I got a r2 score of 72% 
This is a statistical measure of fit that indicates how much variation of a dependent variable is explained by the independent variable.

In [ ]:
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

coefficient_values = model.coef_

df_coefficients = pd.DataFrame(coefficient_values, columns=features).T

df_coefficients.columns = ['coefficient']

df_coefficients

coefficient
Mols KCL     -0.766313
Size nm^3     0.000799

In [ ]:
model.coef_ , model.intercept_

(array([[-0.76631279,  0.00079873]]), array([456.45436197]))

$ h(x) = (\theta_0) + (\theta_1 * x_1 )+ (\theta_2 * x_2) $

$ h(x) = (\theta_0) + (\theta_1 * -0.76631279)  + (\theta_2 * 0.00079873) $

# Part 4. Use Cross Validation

In [ ]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

# model = svm.SVC(kernel='linear', C=1, random_state=0)

scores = cross_val_score(model, x_train, y_train, cv=2)
scores


array([0.68052641, 0.70016562])

In [ ]:
n_samples = x.shape[0]
cv = ShuffleSplit(n_splits = 5, test_size = 0.3, random_state = 0)
cross_val_score(model, x, y, cv = 10)

array([0.66188303, 0.65384327, 0.70364025, 0.64105629, 0.75257321,
       0.71197562, 0.69838721, 0.74670181, 0.64573598, 0.69400968])

In [ ]:
# Report on their finding and their significance

# My values ranges between 0.64 to 0.75 which means that I did not just get lucky with my previous score.

# Part 5. Using Polynomial Regression

In [ ]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

# poly (x_train, x_test)
# fit the model line 36
# write the equation

# x_train
trans = PolynomialFeatures(degree = 2)
trans.fit_transform(x_train)

array([[1.00000000e+00, 6.10000000e+01, 2.51831429e+03, 3.72100000e+03,
        1.53617171e+05, 6.34190684e+06],
       [1.00000000e+00, 3.86000000e+02, 6.51841029e+05, 1.48996000e+05,
        2.51610637e+08, 4.24896727e+11],
       [1.00000000e+00, 7.72000000e+02, 8.23036114e+05, 5.95984000e+05,
        6.35383880e+08, 6.77388445e+11],
       ...,
       [1.00000000e+00, 8.01000000e+02, 2.65273457e+05, 6.41601000e+05,
        2.12484039e+08, 7.03700070e+10],
       [1.00000000e+00, 7.44000000e+02, 1.10481531e+06, 5.53536000e+05,
        8.21982594e+08, 1.22061688e+12],
       [1.00000000e+00, 9.79000000e+02, 9.98594029e+05, 9.58441000e+05,
        9.77623554e+08, 9.97190034e+11]])

In [ ]:
x_poly = trans.transform(x_train)

In [ ]:
model = LinearRegression().fit(x_poly, y_train)

In [ ]:
print('The intercept is:', model.intercept_)

The intercept is: [433.55192854]


In [ ]:
print('The coeffients are:', model.coef_)

The coeffients are: [[ 0.00000000e+00 -1.91851446e+00  2.09486321e-03  1.66478727e-03
  -1.80203746e-06 -2.17871432e-12]]


In [ ]:
# x_test

In [ ]:
trans2 = PolynomialFeatures(degree = 2)
trans2.fit_transform(x_test)

array([[1.00000000e+00, 2.65000000e+02, 1.70568429e+05, 7.02250000e+04,
        4.52006336e+07, 2.90935888e+10],
       [1.00000000e+00, 1.45000000e+02, 2.87802714e+05, 2.10250000e+04,
        4.17313936e+07, 8.28304024e+10],
       [1.00000000e+00, 7.86000000e+02, 1.57789131e+06, 6.17796000e+05,
        1.24022257e+09, 2.48974100e+12],
       [1.00000000e+00, 3.17000000e+02, 4.55071143e+04, 1.00489000e+05,
        1.44257552e+07, 2.07089745e+09],
       [1.00000000e+00, 2.77000000e+02, 3.55862571e+04, 7.67290000e+04,
        9.85739323e+06, 1.26638170e+09],
       [1.00000000e+00, 5.35000000e+02, 3.17629857e+05, 2.86225000e+05,
        1.69931974e+08, 1.00888726e+11],
       [1.00000000e+00, 6.65000000e+02, 3.21295000e+05, 4.42225000e+05,
        2.13661175e+08, 1.03230477e+11],
       [1.00000000e+00, 6.01000000e+02, 1.07378403e+06, 3.61201000e+05,
        6.45344201e+08, 1.15301214e+12],
       [1.00000000e+00, 3.47000000e+02, 5.21542571e+04, 1.20409000e+05,
        1.80975272e+07, 

In [ ]:
x2_poly = trans2.transform(x_test)

model = LinearRegression().fit(x2_poly, y_test)

In [ ]:
print('The intercept is:', model.intercept_)

The intercept is: [556.59522473]


In [ ]:
print('The coeffient is:', model.coef_)

The coeffient is: [[ 0.00000000e+00 -2.50547471e+00  2.14544485e-03  2.42487062e-03
  -2.26388257e-06  1.61148234e-10]]


In [ ]:
# Report on the metrics and output the resultant equation as you did in Part 3.


$ h(x) = (\theta_0) + (\theta_1 * x_1 )+ (\theta_2 * x_2) + (\theta_3 * x_3 )+ (\theta_4 * x_4) + (\theta_5 * x_5 )+ (\theta_6 * x_6) $

# x_train

$ h(x) = (\theta_0) + (\theta_1 * 0.00000000e+00 )+ (\theta_2 * -1.91851446e+00) + (\theta_3 * 2.09486321e-03 )+ (\theta_4 * 1.66478727e-03) + (\theta_5 * -1.80203746e-06 )+ (\theta_6 * -2.17871432e-12) $

# x_test

$ h(x) = (\theta_0) + (\theta_1 * 0.00000000e+00 )+ (\theta_2 * -2.50547471e+00) + (\theta_3 * 2.14544485e-03 )+ (\theta_4 * 2.42487062e-03) + (\theta_5 * -2.26388257e-06 )+ (\theta_6 * 1.61148234e-10) $